Mathlib/Algebra/Ring/Basic.lean

```diff
@@ -174,20 +175,45 @@ lemma IsRightCancelMulZero.to_noZeroDivisors [MulZeroClass α]
   eq_zero_or_eq_zero_of_mul_eq_zero {_ y} h :=
     or_iff_not_imp_right.mpr fun ne ↦ mul_right_cancel₀ ne ((zero_mul y).symm ▸ h)
 
-instance (priority := 100) NoZeroDivisors.to_isCancelMulZero
-    [NonUnitalNonAssocRing α] [NoZeroDivisors α] :
-    IsCancelMulZero α where
-  mul_left_cancel_of_ne_zero ha _ _ h := by
-    rw [← sub_eq_zero, ← mul_sub] at h
-    exact sub_eq_zero.1 ((eq_zero_or_eq_zero_of_mul_eq_zero h).resolve_left ha)
-  mul_right_cancel_of_ne_zero hb _ _ h := by
-    rw [← sub_eq_zero, ← sub_mul] at h
-    exact sub_eq_zero.1 ((eq_zero_or_eq_zero_of_mul_eq_zero h).resolve_right hb)
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] {r : R}
+
+lemma isLeftRegular_iff_right_eq_zero_of_mul : IsLeftRegular r ↔ ∀ x, r * x = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, mul_zero])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [mul_sub, eq, sub_self]
+
+lemma isRightRegular_iff_left_eq_zero_of_mul : IsRightRegular r ↔ ∀ x, x * r = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, zero_mul])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [sub_mul, eq, sub_self]
+
+lemma isRegular_iff_eq_zero_of_mul :
+    IsRegular r ↔ (∀ x, r * x = 0 → x = 0) ∧ (∀ x, x * r = 0 → x = 0) := by
+  rw [isRegular_iff, isLeftRegular_iff_right_eq_zero_of_mul, isRightRegular_iff_left_eq_zero_of_mul]
+
+/-- A (not necessarily unital or associative) ring with no zero divisors has cancellative
+multiplicative on both sides. Since either left or right cancellative multiplication implies
+the absence of zero divisors, the four conditions are equivalent to each other. -/
+lemma noZeroDivisors_tfae : List.TFAE
+    [NoZeroDivisors R, IsLeftCancelMulZero R, IsRightCancelMulZero R, IsCancelMulZero R] := by
+  simp_rw [isLeftCancelMulZero_iff, isRightCancelMulZero_iff, isCancelMulZero_iff_forall_isRegular,
+    isLeftRegular_iff_right_eq_zero_of_mul, isRightRegular_iff_left_eq_zero_of_mul,
+    isRegular_iff_eq_zero_of_mul]
+  tfae_have 1 ↔ 2 := noZeroDivisors_iff_right_eq_zero_of_mul
+  tfae_have 1 ↔ 3 := noZeroDivisors_iff_left_eq_zero_of_mul
+  tfae_have 1 ↔ 4 := noZeroDivisors_iff_eq_zero_of_mul
+  tfae_finish
 
 /-- In a ring, `IsCancelMulZero` and `NoZeroDivisors` are equivalent. -/
-lemma isCancelMulZero_iff_noZeroDivisors [NonUnitalNonAssocRing α] :
-    IsCancelMulZero α ↔ NoZeroDivisors α :=
-  ⟨fun _ => IsRightCancelMulZero.to_noZeroDivisors _, fun _ => inferInstance⟩
+lemma isCancelMulZero_iff_noZeroDivisors : IsCancelMulZero R ↔ NoZeroDivisors R :=
+  noZeroDivisors_tfae.out 3 0
+
+variable (R) in
+instance (priority := 100) NoZeroDivisors.to_isCancelMulZero
+    [NoZeroDivisors R] : IsCancelMulZero R :=
+  isCancelMulZero_iff_noZeroDivisors.mpr ‹_›
```

Comment:
Is this a good instance? `NoZeroDivisors R` always implies `IsCancelMulZero R` while the other way around only works for rings, and therefore I would think we want to push people towards using `IsCancelMulZero` over `NoZeroDivisors`

Reply:
Yeah but this is also pre-existing and I merely moved it. I'd try downgrade this to a theorem and fix any breakage in a separate PR.
I think you mean that `IsCancelMulZero` always implies `NoZeroDivisors`, though it still need a `MulZeroClass` assumption, which is of course milder than `NonUnitalNonAssocRing`. I mentioned above that I'd like to make this an instance, but also in a separate PR.

Reply:
For now you can look at the declarations diff; I don't know where the `+ _` come from ...
<img width="516" height="759" alt="image" src="https://github.com/user-attachments/assets/c0276fc2-3753-485b-96fd-56680cb90c66" />

Reply:
Yes sorry I am of course aware of the declaration diff, but I mean a more inline tool. Maybe a tampermonkey extension?